### ISSUE TITLE
Uncaught TypeError: Cannot read property  slice  of undefined

### ISSUE THREAD

**Issue by Talanor** [atom/atom#4619]
### Steps To Reproduce:
- Create an __ init __.py in a python package
- Try to add a few new lines by typing enter (I can't have more than two lines)
### Error

Uncaught TypeError: Cannot read property  slice  of undefined **Atom Version**: 0.158.0 **System**: linux 3.13.0-43-generic **Thrown From**: Atom Core
### Stack Trace

```
At /usr/share/atom/resources/app/node_modules/text-buffer/lib/text-buffer.js:557

TypeError: Cannot read property 'slice' of undefined
  at TextBuffer.module.exports.TextBuffer.applyPatch (/usr/share/atom/resources/app/node_modules/text-buffer/lib/text-buffer.js:557:39)
  at BufferPatch.module.exports.BufferPatch.applyTo (/usr/share/atom/resources/app/node_modules/text-buffer/lib/buffer-patch.js:74:21)
  at Transaction.module.exports.Transaction.applyTo (/usr/share/atom/resources/app/node_modules/text-buffer/lib/transaction.js:60:29)
  at History.module.exports.History.abortTransaction (/usr/share/atom/resources/app/node_modules/text-buffer/lib/history.js:162:24)
  at History.module.exports.History.transact (/usr/share/atom/resources/app/node_modules/text-buffer/lib/history.js:119:16)
  at TextBuffer.module.exports.TextBuffer.transact (/usr/share/atom/resources/app/node_modules/text-buffer/lib/text-buffer.js:683:27)
  at TextEditor.module.exports.TextEditor.transact (/usr/share/atom/resources/app/src/text-editor.js:1146:26)
  at atom-text-editor.newCommandListeners.(anonymous function) (/usr/share/atom/resources/app/src/text-editor-element.js:290:22)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/usr/share/atom/resources/app/src/command-registry.js:243:29)
  at /usr/share/atom/resources/app/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/usr/share/atom/resources/app/node_modules/atom-keymap/lib/keymap-manager.js:549:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/usr/share/atom/resources/app/node_modules/atom-keymap/lib/keymap-manager.js:391:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/usr/share/atom/resources/app/src/window-event-handler.js:167:20)
```

**Comment by 50Wliu** [comment]
#4749 and #5912 have both referenced undoing.

**Comment by ChallahuAkbar** [comment]
I just had this issue when trying to redo, here's the stack trace:

```
At /opt/atom/resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:682

TypeError: Cannot read property 'slice' of undefined
    at TextBuffer.module.exports.TextBuffer.applyChange (/opt/atom/resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:682:41)
    at TextBuffer.module.exports.TextBuffer.redo (/opt/atom/resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:848:16)
    at /opt/atom/resources/app.asar/src/text-editor.js:1395:31
    at TextEditor.module.exports.TextEditor.mergeSelections (/opt/atom/resources/app.asar/src/text-editor.js:2381:18)
    at TextEditor.module.exports.TextEditor.avoidMergingSelections (/opt/atom/resources/app.asar/src/text-editor.js:2363:35)
    at TextEditor.module.exports.TextEditor.redo (/opt/atom/resources/app.asar/src/text-editor.js:1393:12)
    at TextEditor.commandRegistry.add.stopEventPropagation.core:redo (/opt/atom/resources/app.asar/src/register-default-commands.js:305:21)
    at atom-text-editor.newCommandListeners.(anonymous function) (/opt/atom/resources/app.asar/src/register-default-commands.js:642:32)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/opt/atom/resources/app.asar/src/command-registry.js:260:29)
    at /opt/atom/resources/app.asar/src/command-registry.js:3:61
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/opt/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:580:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/opt/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:388:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/opt/atom/resources/app.asar/src/window-event-handler.js:98:36)
    at HTMLDocument.<anonymous> (/opt/atom/resources/app.asar/src/window-event-handler.js:3:61)
```

**Comment by entozoon** [comment]
Same for me here. very similar (if not same) stack trace as above:
```
Uncaught TypeError: Cannot read property 'slice' of undefined
C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\node_modules\text-buffer\lib\text-buffer.js:633
Hide Stack Trace
TypeError: Cannot read property 'slice' of undefined
    at TextBuffer.module.exports.TextBuffer.applyChange (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\node_modules\text-buffer\lib\text-buffer.js:633:39)
    at TextBuffer.module.exports.TextBuffer.undo (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\node_modules\text-buffer\lib\text-buffer.js:782:16)
    at C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\text-editor.js:1532:31
    at TextEditor.module.exports.TextEditor.mergeSelections (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\text-editor.js:2522:43)
    at TextEditor.module.exports.TextEditor.avoidMergingSelections (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\text-editor.js:2508:35)
    at TextEditor.module.exports.TextEditor.undo (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\text-editor.js:1530:12)
    at TextEditor.core:undo (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\register-default-commands.js:307:21)
    at atom-text-editor.newCommandListeners.(anonymous function) (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\register-default-commands.js:647:32)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\command-registry.js:259:29)
    at C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\command-registry.js:3:59
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\node_modules\atom-keymap\lib\keymap-manager.js:599:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\node_modules\atom-keymap\lib\keymap-manager.js:390:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\window-event-handler.js:106:36)
    at HTMLDocument.<anonymous> (C:\Users\Myself\AppData\Local\atom\app-1.14.0-beta1\resources\app.asar\src\window-event-handler.js:3:59)
```
Atom 1.14.0-beta1 ia32, Windows 7.

Steps to re-produce it (consistently!):
- Open a file (.scss in my case), 
- Paste some new text over its contents
- Move a line up or down
- Save it
- Ctrl-Z a couple times.

**Comment by Ben3eeE** [comment]
@entozoon Can you reproduce it when you start Atom in safe mode `atom --safe`? It would also be great if you could share the files and text you are pasting that reproduces this consistently, thanks.

**Comment by entozoon** [comment]
Yep, reproduced in `atom --safe`, sometimes, although it is intermittent.

_footer.scss ::
```
.footer {
	position: relative;
	background: red;
	color: red;
}
```
- Create and open that file in `atom --safe`
- Ctrl-A and Ctrl-V to paste the following over the top:
```
.footer {
	@include foo;
	.bar {
		margin-left: 10px;
	}
}
```
- Move the line `margin-left: 10px;` up above `@include` (Ctrl-Up twice)
- Ctrl-S save
- Ctrl-Z undo a few times


Here is a simpler situation where a similar problem occurs, and it feels like the same issue:
- Create and open that file again in `atom --safe`
- Move the `color: red;` line two lines up (ctrl-up twice)
- Ctrl-S save
- Ctrl-Z undo

And the line disappears!

So I get the impression it's something to do with lines containing @import, although I can't imagine why :)

**Comment by Ben3eeE** [comment]
@entozoon Thank you for the additional information! With this information I was able to reproduce the issue.

Here is what I did to consistently reproduce in safe mode:

1. Copy the contents of this file into a new untitled tab.
```scss
.footer {
	position: relative;
	background: red;
	color: red;
}
```
2. Place the cursor on the 4th line
1. Quicky press <kbd>Ctrl-Up</kbd> twice to move the line up
1. Quickly press <kbd>Ctrl-z</kbd> the line disappears
1. Press <kbd>Ctrl-z</kbd> again
1.  :boom: that links to this issue

**Gif of Reproduction:**
![splice of undefined](https://cloud.githubusercontent.com/assets/1058982/22019656/c67eba4a-dcb4-11e6-8994-1271d09b2830.gif)


Reproduced with `Atom 1.15.0-dev-1592b1e x64`  on Windows 7.

**Comment by nathansobo** [comment]
@Ben3eeE Is this a regression in 1.14?

**Comment by nathansobo** [comment]
Investigating this today. It seems like the exception is due to referencing a non-existent row when applying the change during the undo operation. Probably a downstream effect of that line disappearing. Stay tuned.

**Comment by nathansobo** [comment]
I've found the root cause and have a [PR open](https://github.com/atom/superstring/pull/8) on the native Patch component to fix an oversight in the implementation. We should have a hotfix out soon for this issue.

**Comment by nathansobo** [comment]
This should be resolved via 1.14.0-beta2 which was just published.
